### ISSUE TITLE
support of audioread/librosa package

### ISSUE THREAD

**Issue by rjt10** [chaquo/chaquopy#21]
Got error when trying to pip install these two packages with this:
```
            pip {
                install "six==1.10.0"
                install "audioread==2.1.5"
            }
```
The error message is: 
```
Collecting audioread==2.1.5

  Could not find a version that satisfies the requirement audioread==2.1.5 (from versions: )
No matching distribution found for audioread==2.1.5 (NOTE: Chaquopy only supports wheels, not sdist packages)
Exit status 
```

Looking at the past feature requests, I have an impression that the python packages need to be manually ported to be available through Chaquopy, it is the case? If so, I am wondering if there's a way for community to help porting them?

Thanks.

**Comment by mhsmith** [comment]
Packages which are available on PyPI as pure Python wheels don't need to be rebuilt, but all others do. We have some unreleased scripts to assist in this, which work fine for packages which are pure Python or self-contained native code. However, we can't currently handle packages which depend on external native libraries or executables. In order to use these on Android, we'd also need a procedure for building and packaging the external components themselves.

This is the case with audioread, which (like many Python audio decoding libraries) depends on FFmpeg. If you only need to read uncompressed audio files, it looks like it'll still work, in which case I can add it to our repository if you like. Otherwise, you'll have to use a different approach. The obvious way would be to use Chaquopy to access the Android MediaCodec API from Python, though that looks pretty unfriendly.

As for librosa, we unfortunately can't yet build that because it depends on SciPy, which is difficult for other reasons (#15). The current version also depends on LLVM, though it looks this could be avoided by using an older version.

We are definitely considering making our package build system open-source so the community can help with things like this. I'll update this issue if that happens.

**Comment by liangyue62** [comment]
Sorry for asking, but I wonder is librosa available now?
Because I have seen that SciPy is now available. Or is there still have other problems?

**Comment by mhsmith** [comment]
The current version of librosa (0.6.1) still can't be installed because it depends on numba and llvmlite. However, you should be able to install the older version 0.4.2, which doesn't have those dependencies. I'm not familiar with librosa so I haven't tested whether it works: please give it a try and report back.

**Comment by lorenoooo** [comment]
Hi, 
I get an error even using the older vesion 0.4.2 of librosa.

I include ```classpath "com.chaquo.python:gradle:5.0.0"``` on the project graddle.

This is my app graddle

```
apply plugin: 'com.android.application'
apply plugin: 'com.chaquo.python'

android {
    compileSdkVersion 27
    buildToolsVersion '28.0.3'
    defaultConfig {
        applicationId [...]
        minSdkVersion 21
        targetSdkVersion 26
        compileSdkVersion 27
        versionCode 1
        versionName "1.0"
        ndk{
            abiFilters "armeabi-v7a"
        }
        python {
            buildPython "C:/Users/[...]/AppData/Local/Programs/Python/Python35/python.exe"
            pip {
                //install "librosa==0.4.2"
                //install "-r", "requeriments.txt"
                install "https://github.com/librosa/librosa/archive/0.4.2.zip"
            }
        }
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
    productFlavors {
    }
    repositories {
        maven {
            url 'https://google.bintray.com/tensorflow'
        }
    }
    aaptOptions {
        noCompress "tflite"
        noCompress "lite"
    }
}

dependencies {
    [...]
}
```

I get the error after sucessfully built librosa:
```
Successfully built librosa
Installing collected packages: audioread, chaquopy-libgfortran, chaquopy-openblas, numpy, chaquopy-gnustl, scipy, scikit-learn, chaquopy-libpng, six, pyparsing, chaquopy-freetype, pytz, setuptools, kiwisolver, python-dateutil, cycler, matplotlib, joblib, decorator, librosa
Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: 'C:\\Users\\[...]\\AppData\\Local\\Programs\\Python\\Python35\\Lib\\site-packages\\tensorflow\\examples\\android\\LULLAAI\\app\\build\\generated\\python\\requirements\\debug\\armeabi-v7a\\matplotlib\\backends\\web_backend\\jquery\\css\\themes\\base\\images\\ui-bg_diagonals-thick_18_b81900_40x40.png'

Chaquopy: Exit status 1
:app:generateDebugPythonRequirements FAILED
```

It's a little bit strange. I cleaned the project. I restarted android studio as administrator (I'm developing my app in windows)

Can you help me ? Must I open a new issue?

Thanks

**Comment by mhsmith** [comment]
This is probably caused by the Windows filename length limit of 260 characters. Try moving your project to a location with a shorter path.

**Comment by lorenoooo** [comment]
Yes, this was the problem. I didn't realized. Thanks!

Now, I get a new error, but I think it is a problem with python.
´´´
if (!Python.isStarted()) {
	Python.start(new AndroidPlatform(context));
}

py = Python.getInstance();
librosa = py.getModule("librosa");
 ´´´
When I make a cool to get the librosa module, an exception is thrown:
´´´
com.chaquo.python.PyException: ImportError: This platform lacks a functioning sem_open implementation, therefore, the required synchronization primitives needed will not function, see issue 3770.
´´´

Do i make an incorrect call to import the librosa lib, or is a problem of my python instalation? In latter, can you say me where I must ask for them? I can't find a solution googling. Thanks in advance.

**Comment by mhsmith** [comment]
Full exception: 

```
>>> import librosa
/android_asset/chaquopy/requirements-common.zip/joblib/_multiprocessing_helpers.py:38: UserWarning: This platform lacks a functioning sem_open implementation, therefore, the required synchronization primitives needed will not function, see issue 3770..  joblib will operate in serial mode
Traceback (most recent call last):
  File "stdlib/multiprocessing/synchronize.py", line 29, in <module>
ImportError: cannot import name 'sem_unlink'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "import.pxi", line 59, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/_vendor/six.py", line 686, in reraise
    raise value
  File "import.pxi", line 18, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 331, in load_module
    self.load_module_impl()
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 411, in load_module_impl
    six.exec_(self.get_code(self.mod_name), mod.__dict__)
  File "/android_asset/chaquopy/requirements-common.zip/librosa/__init__.py", line 14, in <module>
    from . import cache
  File "import.pxi", line 59, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/_vendor/six.py", line 686, in reraise
    raise value
  File "import.pxi", line 18, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 331, in load_module
    self.load_module_impl()
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 411, in load_module_impl
    six.exec_(self.get_code(self.mod_name), mod.__dict__)
  File "/android_asset/chaquopy/requirements-common.zip/librosa/cache.py", line 7, in <module>
    from joblib import Memory
  File "import.pxi", line 59, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/_vendor/six.py", line 686, in reraise
    raise value
  File "import.pxi", line 18, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 331, in load_module
    self.load_module_impl()
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 411, in load_module_impl
    six.exec_(self.get_code(self.mod_name), mod.__dict__)
  File "/android_asset/chaquopy/requirements-common.zip/joblib/__init__.py", line 119, in <module>
    from .parallel import Parallel
  File "import.pxi", line 59, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/_vendor/six.py", line 686, in reraise
    raise value
  File "import.pxi", line 18, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 331, in load_module
    self.load_module_impl()
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 411, in load_module_impl
    six.exec_(self.get_code(self.mod_name), mod.__dict__)
  File "/android_asset/chaquopy/requirements-common.zip/joblib/parallel.py", line 33, in <module>
    from .externals import loky
  File "import.pxi", line 59, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/_vendor/six.py", line 686, in reraise
    raise value
  File "import.pxi", line 18, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 331, in load_module
    self.load_module_impl()
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 411, in load_module_impl
    six.exec_(self.get_code(self.mod_name), mod.__dict__)
  File "/android_asset/chaquopy/requirements-common.zip/joblib/externals/loky/__init__.py", line 11, in <module>
    from .backend.context import cpu_count
  File "import.pxi", line 59, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/_vendor/six.py", line 686, in reraise
    raise value
  File "import.pxi", line 18, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 331, in load_module
    self.load_module_impl()
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/android/importer.py", line 411, in load_module_impl
    six.exec_(self.get_code(self.mod_name), mod.__dict__)
  File "/android_asset/chaquopy/requirements-common.zip/joblib/externals/loky/backend/__init__.py", line 13, in <module>
    from multiprocessing import synchronize
  File "import.pxi", line 59, in java.chaquopy.import_override
  File "/data/user/0/com.chaquo.python.pkgtest3/files/chaquopy/bootstrap.zip/java/_vendor/six.py", line 686, in reraise
    raise value
  File "import.pxi", line 18, in java.chaquopy.import_override
  File "stdlib/multiprocessing/synchronize.py", line 34, in <module>
ImportError: This platform lacks a functioning sem_open implementation, therefore, the required synchronization primitives needed will not function, see issue 3770.
```

**Comment by mhsmith** [comment]
Despite the message, it appears that the current version of joblib (0.13.0) does not actually manage to fall back to serial mode.

Try inserting this line in `build.gradle`, *before* the librosa line:
```
install "joblib==0.11"
```

When I did that, I still received some warnings, but the librosa import completed successfully.

**Comment by lorenoooo** [comment]
Thanks, give some days to check it. On Monday i'll give you some response

**Comment by lorenoooo** [comment]
Finally, I found the librosa MFCC features extraction programmed on Java, which were the functionality I needed from librosa in my Android app. So, I didn't test whether your last solution could solve my problem. Thanks a lot, I hope it will resolve the problem to other programmers!

**Comment by mhsmith** [comment]
This issue is fixed in Chaquopy 7.0.2. To upgrade, edit your app's top-level build.gradle file and change the version number of `com.chaquo.python:gradle` to `7.0.2`.

Please review the [release notes](https://chaquo.com/chaquopy/doc/current/changelog.html), especially the point about most native packages having been upgraded.

**Comment by Neur1n** [comment]
Hello, I've tried to use Chaquopy 7.0.2 and 7.0.3 but still couldn't get it works. A `ModuleNotFoundError: No module named 'librosa'` exception was thrown. More information:

- OS: Windows 10
- librosa version: 0.6.1 ('minimal' version that supports joblib of 0.11.0)
- joblib version: 0.11.0

**Comment by mhsmith** [comment]
Please post:
* The full exception trace
* The section of source code which caused it
* The `python` section of your build.gradle file

By the way, the note in #185 is out of date: you can now use any version of librosa. However, you may still need to use joblib 0.11.0, or the workaround at https://github.com/joblib/joblib/issues/825#issuecomment-506762348.

**Comment by Neur1n** [comment]
## Exception Trace
   --------- beginning of crash
2020-06-04 11:29:00.016 3027-3027/<PROJECT_NAME> E/AndroidRuntime: FATAL EXCEPTION: main
    Process: <PROJECT_NAME>, PID: 3027
    java.lang.RuntimeException: Unable to start activity ComponentInfo{<PROJECT_NAME>/<PROJECT_NAME>.MainActivity}: com.chaquo.python.PyException: ModuleNotFoundError: No module named 'librosa'
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: com.chaquo.python.PyException: ModuleNotFoundError: No module named 'librosa'
        at <python>.java.chaquopy.import_override(import.pxi:20)
        at <python>.java.chaquopy.import_override(import.pxi:60)
        at <python>.main.<module>(main.py:5)
        at <python>.importlib._bootstrap._call_with_frames_removed(<frozen importlib._bootstrap>:219)
        at <python>.importlib._bootstrap_external.exec_module(<frozen importlib._bootstrap_external>:783)
        at <python>.java.android.importer.exec_module(importer.py:503)
        at <python>.java.android.importer.exec_module(importer.py:546)
        at <python>.importlib._bootstrap._load_unlocked(<frozen importlib._bootstrap>:671)
        at <python>.importlib._bootstrap._find_and_load_unlocked(<frozen importlib._bootstrap>:975)
        at <python>.importlib._bootstrap._find_and_load(<frozen importlib._bootstrap>:991)
        at <python>.importlib._bootstrap._gcd_import(<frozen importlib._bootstrap>:1014)
        at <python>.importlib.import_module(__init__.py:127)
        at <python>.chaquopy_java.Java_com_chaquo_python_Python_getModule(chaquopy_java.pyx:152)
        at com.chaquo.python.Python.getModule(Native Method)
        at <PROJECT_NAME>.MainActivity.onStart(MainActivity.java:42)
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1334)
        at android.app.Activity.performStart(Activity.java:7056)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2741)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 

## Source Code
### main.py
```python
import librosa
def test():
    print(librosa.__version__)
```
### MainActivity.java
```java
@Override
protected void onStart() {
    super.onStart()
    if (!Python.isStarted()) {
        Python.start(new AndroidPlatform(this));
    }
    Python py = Python.getInstance();
    py.getModule("main").callAttr("test");
}
```

## Gradle
### Project Gradle
```groovy
buildscript {
    repositories {
        maven { url "https://chaquo.com/maven" }
    }
    dependencies {
        classpath "com.chaquo.python:gradle:7.0.3"
    }
}
```
### Module Gradle
```groovy
apply plugin: 'com.android.application'
apply plugin: 'com.chaquo.python'

android {
        ndk {
            abiFilters "armeabi-v7a"
        }
    }

    sourceSets {
        main {
            python {
                srcDir "src/main/python"
            }
        }
}
```

**Comment by Neur1n** [comment]
I just realize what does "The python section of your build.gradle file" mean, and I did miss this section. Now everything works.

**Comment by mhsmith** [comment]
I've just learned that although the current version of librosa can be installed with Chaquopy, it still fails to import. This will be fixed in the next version of Chaquopy.

**Comment by mhsmith** [comment]
This issue is fixed in Chaquopy 8.0.1. If you're using that version or later, there's no longer any need to use old versions of librosa or joblib.

To upgrade, edit your app's top-level build.gradle file and change the version number of `com.chaquo.python:gradle`.
